File: runtimes/net/Sources/Parsers/BaseLRParser.cs
/**********************************************************************
* Copyright (c) 2013 Laurent Wouters and others
* This program is free software: you can redistribute it and/or modify
* it under the terms of the GNU Lesser General Public License as
* published by the Free Software Foundation, either version 3
* of the License, or (at your option) any later version.
*
* This program is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
* GNU Lesser General Public License for more details.
*
* You should have received a copy of the GNU Lesser General
* Public License along with this program.
* If not, see <http://www.gnu.org/licenses/>.
*
* Contributors:
*     Laurent Wouters - lwouters@xowl.org
**********************************************************************/
using System.Collections.Generic;
using Hime.Redist.Utils;

namespace Hime.Redist.Parsers
{
	/// <summary>
	/// Represents a base LR parser
	/// </summary>
	public abstract class BaseLRParser
	{
		/// <summary>
		/// Maximum number of errors
		/// </summary>
		protected const int MAX_ERROR_COUNT = 100;
		/// <summary>
		/// The default value of the recover mode
		/// </summary>
		protected const bool DEFAULT_MODE_RECOVER = true;
		/// <summary>
		/// The default value of the debug mode
		/// </summary>
		protected const bool DEFAULT_MODE_DEBUG = false;

		/// <summary>
		/// Parser's variable symbols
		/// </summary>
		protected readonly ROList<Symbol> symVariables;
		/// <summary>
		/// Parser's virtual symbols
		/// </summary>
		protected readonly ROList<Symbol> symVirtuals;
		/// <summary>
		/// Parser's action symbols
		/// </summary>
		protected readonly ROList<SemanticAction> symActions;
		/// <summary>
		/// List of the encountered syntaxic errors
		/// </summary>
		protected readonly List<ParseError> allErrors;
		/// <summary>
		/// Lexer associated to this parser
		/// </summary>
		protected readonly Lexer.BaseLexer lexer;


		/// <summary>
		/// Gets the variable symbols used by this parser
		/// </summary>
		public ROList<Symbol> SymbolVariables { get { return symVariables; } }

		/// <summary>
		/// Gets the virtual symbols used by this parser
		/// </summary>
		public ROList<Symbol> SymbolVirtuals { get { return symVirtuals; } }

		/// <summary>
		/// Gets the action symbols used by this parser
		/// </summary>
		public ROList<SemanticAction> SymbolActions { get { return symActions; } }

		/// <summary>
		/// Gets or sets whether the paser should try to recover from errors
		/// </summary>
		public bool ModeRecoverErrors
		{
			get;
			set;
		}

		/// <summary>
		/// Gets or sets a value indicating whether this parser is in debug mode
		/// </summary>
		public bool ModeDebug
		{
			get;
			set;
		}

		/// <summary>
		/// Initializes a new instance of the LRkParser class with the given lexer
		/// </summary>
		/// <param name="variables">The parser's variables</param>
		/// <param name="virtuals">The parser's virtuals</param>
		/// <param name="actions">The parser's actions</param>
		/// <param name="lexer">The input lexer</param>
		protected BaseLRParser(Symbol[] variables, Symbol[] virtuals, SemanticAction[] actions, Lexer.BaseLexer lexer)
		{
			ModeRecoverErrors = DEFAULT_MODE_RECOVER;
			ModeDebug = DEFAULT_MODE_DEBUG;
			symVariables = new ROList<Symbol>(new List<Symbol>(variables));
			symVirtuals = new ROList<Symbol>(new List<Symbol>(virtuals));
			symActions = new ROList<SemanticAction>(actions != null ? new List<SemanticAction>(actions) : new List<SemanticAction>());
			ModeRecoverErrors = true;
			allErrors = new List<ParseError>();
			this.lexer = lexer;
			this.lexer.OnError += OnLexicalError;
		}

		/// <summary>
		/// Adds the given lexical error emanating from the lexer to the list of errors
		/// </summary>
		/// <param name="error">Lexical error</param>
		protected void OnLexicalError(ParseError error)
		{
			allErrors.Add(error);
		}

		/// <summary>
		/// Parses the input and returns the result
		/// </summary>
		/// <returns>A ParseResult object containing the data about the result</returns>
		public abstract ParseResult Parse();
	}
}

File: runtimes/java/src/fr/cenotelie/hime/redist/parsers/BaseLRParser.java
/*******************************************************************************
 * Copyright (c) 2017 Association Cénotélie (cenotelie.fr)
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as
 * published by the Free Software Foundation, either version 3
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General
 * Public License along with this program.
 * If not, see <http://www.gnu.org/licenses/>.
 ******************************************************************************/
package fr.cenotelie.hime.redist.parsers;

import fr.cenotelie.hime.redist.ParseError;
import fr.cenotelie.hime.redist.ParseResult;
import fr.cenotelie.hime.redist.SemanticAction;
import fr.cenotelie.hime.redist.Symbol;
import fr.cenotelie.hime.redist.lexer.BaseLexer;
import fr.cenotelie.hime.redist.lexer.LexicalErrorHandler;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;

/**
 * Represents a base LR parser
 *
 * @author Laurent Wouters
 */
public abstract class BaseLRParser {
    /**
     * Maximum number of errors
     */
    static final int MAX_ERROR_COUNT = 100;
    /**
     * The default value of the recover mode
     */
    private static final boolean DEFAULT_MODE_RECOVER = true;
    /**
     * The default value of the debug mode
     */
    private static final boolean DEFAULT_MODE_DEBUG = false;

    /**
     * Determines whether the parser will try to recover from errors
     */
    private boolean modeRecover;
    /**
     * Value indicating whether this parser is in debug mode
     */
    private boolean modeDebug;
    /**
     * Parser's variables
     */
    final List<Symbol> symVariables;
    /**
     * Parser's virtuals
     */
    final List<Symbol> symVirtuals;
    /**
     * Parser's actions
     */
    final List<SemanticAction> symActions;
    /**
     * List of the encountered syntaxic errors
     */
    final List<ParseError> allErrors;
    /**
     * Lexer associated to this parser
     */
    final BaseLexer lexer;


    /**
     * Gets the variable symbols used by this parser
     *
     * @return The variable symbols used by this parser
     */
    public List<Symbol> getVariables() {
        return symVariables;
    }

    /**
     * Gets the virtual symbols used by this parser
     *
     * @return The virtual symbols used by this parser
     */
    public List<Symbol> getVirtuals() {
        return symVirtuals;
    }

    /**
     * Gets whether the parser should try to recover from errors
     *
     * @return Whether the parser should try to recover from errors
     */
    public boolean getModeRecoverErrors() {
        return modeRecover;
    }

    /**
     * Sets whether the parser should try to recover from errors
     *
     * @param recover Whether the parser should try to recover from errors
     */
    public void setModeRecoverErrors(boolean recover) {
        this.modeRecover = recover;
    }

    /**
     * Gets whether this parser is in debug mode
     *
     * @return Whether this parser is in debug mode
     */
    public boolean isModeDebug() {
        return modeDebug;
    }

    /**
     * Sets whether this parser is in debug mode
     *
     * @param mode Whether this parser is in debug mode
     */
    public void setModeDebug(boolean mode) {
        this.modeDebug = mode;
    }

    /**
     * Initializes a new instance of the LRkParser class with the given lexer
     *
     * @param variables The parser's variable
     * @param virtuals  The parser's virtuals
     * @param actions   The parser's actions
     * @param lexer     The input lexer
     */
    protected BaseLRParser(Symbol[] variables, Symbol[] virtuals, SemanticAction[] actions, BaseLexer lexer) {
        this.modeRecover = DEFAULT_MODE_RECOVER;
        this.modeDebug = DEFAULT_MODE_DEBUG;
        this.symVariables = Collections.unmodifiableList(Arrays.asList(variables));
        this.symVirtuals = Collections.unmodifiableList(Arrays.asList(virtuals));
        this.symActions = Collections.unmodifiableList(actions != null ? Arrays.asList(actions) : new ArrayList<SemanticAction>());
        this.allErrors = new ArrayList<>();
        this.lexer = lexer;
        this.lexer.setErrorHandler(new LexicalErrorHandler() {
            @Override
            public void handle(ParseError error) {
                allErrors.add(error);
            }
        });
    }

    /**
     * Parses the input and returns the result
     *
     * @return A ParseResult object containing the data about the result
     */
    public abstract ParseResult parse();
}
